### PR TITLE
Update etcd client to 3.3 for 1.13

### PR DESCRIPTION
**Pull request by jpbetz**
Vendor in the etcd 3.3 client targeting kubernetes 1.13 release milestone. We had previously tested the etcd 3.3 client with kubernetes via https://github.com/kubernetes/kubernetes/pull/58551.

Most of the work here was due to the need to bump to `gogo/protobuf` `v0.5` since the etcd client includes protobuf bindings generated for this version of `gogo/protobuf` and is not compatible with the `v0.4-3-gc0656edd` version previously in use by Kubernetes.  Since we're at the very beginning of the k8s 1.13 release cycle, and we are already lagging behind on `gogo/protobuf` updates, and the `v0.5` version includes noteworthy improvements, like the elimination of unsafe code in the generated bindings, this seems like a good time do the bump and regen.

Also, added `':(exclude)**.*.pb.go'` to `hack/verify-pkg-names.sh` to resolve this lint error:

```
staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1/generated.pb.go:import encoding_binary "encoding/binary"
```

/area etcd
/sig api-machinery
/kind feature
/cc @gyuho @xiang90 @wenjiaswe @jingyih 

```release-note
Upgrade to etcd 3.3 client
```

### CODE REVIEW

**Comment by jpbetz** [comment]
Sorting out some gogo/protobuf + generated code issues, hang tight.

**Comment by gyuho** [comment]
etcd part LGTM.

Just for reference, this does not change clientv3 behavior since we backported health check feature to https://github.com/etcd-io/etcd/blob/master/CHANGELOG-3.2.md#v3210-2017-11-16 client. In addition, this replaces `golang.org/x/net/context` to `context`, which should not break anything.

**Comment by jpbetz** [comment]
This is ready for review.

**Comment by jpbetz** [comment]
cc @timothysc

**Comment by jpbetz** [comment]
cc @lavalamp @sttts Could I get a review? This is mostly dependency bumps with only minimal code change. But it touches a lot of files so if it sits for too long I end up having to do more tedious rebases...

**Comment by lavalamp** [comment]
I looked at this, that's an enormous vendoring change.

Is there any way to confine it to etcd only?

Changing every protobuf file in the repo is making me very nervous.

**Comment by jpbetz** [comment]
> I looked at this, that's an enormous vendoring change.
> 
> Is there any way to confine it to etcd only?

I was careful to the transitive dependency bumps to ones that are actually required.  Without the gogo/protobuf bump, we had a incompatibility in the etcd client:

```
vendor/github.com/coreos/etcd/client/keys.generated.go:70:6: r.WriteArrayStart undefined (type codec.encDriver has no field or method WriteArrayStart)
```

> Changing every protobuf file in the repo is making me very nervous.

Yeah, I wasn't thrilled to find that we needed to regen our protobuf files. But we are lagging behind on gogo/protobuf and this bumps us up from `v0.4-3-gc0656edd` to `v0.5`, which is an officially released version, and eliminates unsafe code from the generated code, see https://github.com/gogo/protobuf/releases.

And we're doing this at the very beginning of the 1.13 cycle, which gives us ample time to soak these changes.

File: cluster/images/etcd/migrate/rollback_v2.go [hunk 1]
```diff
@@ -314,12 +314,33 @@ func toTTLOptions(r *pb.Request) store.TTLOptionSet {
 }
 
 func applyRequest(r *pb.Request, applyV2 etcdserver.ApplierV2) {
+	// TODO: find a sane way to perform this cast or avoid it in the first place
+	reqV2 := &etcdserver.RequestV2{
```
Comment (timothysc) [inline cluster/images/etcd/migrate/rollback_v2.go]: This seems orthogonal to the main purpose of this PR which is level up the deps to 3.3

**Review by timothysc — COMMENTED** [review]
Generally LGTM, but I would split out your v2 compat changes into it's own PR, unless this update somehow forces the issue. 

@lavalamp - This is normal.  Every time we levelset protobuf changes we need to regenerate * which cascades through machinery.  This has happened a number of times. 

/assign @liggitt @wojtek-t  
/approve

File: cluster/images/etcd/migrate/rollback_v2.go (hunk 1 above)
Comment (liggitt) [reply cluster/images/etcd/migrate/rollback_v2.go]: the etcd2-specific tools are being removed in https://github.com/kubernetes/kubernetes/pull/69577, fyi

Comment (jpbetz) [reply cluster/images/etcd/migrate/rollback_v2.go]: Unfortunately, This change is required by the version bump, due to a change in the etcd codebase.  Look forward to stripping out the etcd2 code in 1.14

Comment (timothysc) [reply cluster/images/etcd/migrate/rollback_v2.go]: > Generally LGTM, but I would split out your v2 compat changes into it's own PR, unless this update somehow forces the issue. 

^ seems to apply then.

File: pkg/kubelet/apis/deviceplugin/v1alpha/api.pb.go [hunk 2]
```diff
@@ -820,24 +819,6 @@ func (m *DeviceSpec) MarshalTo(dAtA []byte) (int, error) {
 	return i, nil
 }
 
-func encodeFixed64Api(dAtA []byte, offset int, v uint64) int {
```
Comment (liggitt) [inline pkg/kubelet/apis/deviceplugin/v1alpha/api.pb.go]: do we have any tests that ensure previously serialized data decodes correctly and changes like this stay wire-compatible?

Comment (jpbetz) [reply pkg/kubelet/apis/deviceplugin/v1alpha/api.pb.go]: We have https://k8s-testgrid.appspot.com/sig-release-master-upgrade which I believe are all post-commit. That should be sufficient, right?

Comment (liggitt) [reply pkg/kubelet/apis/deviceplugin/v1alpha/api.pb.go]: that's good in one direction, but not exhaustive for all fields (we have a few oddities in our proto generation around nullable type aliases), and doesn't test in the opposite direction (new serialized data is readable by old servers).  I don't see anything concerning in this PR, but it's pretty impossible to eyeball. @smarterclayton, any specific areas we should look at carefully?

Comment (jpbetz) [reply pkg/kubelet/apis/deviceplugin/v1alpha/api.pb.go]: There are also downgrade tests (https://k8s-testgrid.appspot.com/sig-release-master-upgrade#gce-master-new-downgrade-cluster) that cover the other direction.

But it's certainly not exhaustive. I'm all in favor of capturing the test coverage gap we think we have here and contributing to get it fixed. Should we open a new issue for that? Should we block this PR?

^ @lavalamp

Comment (lavalamp) [reply pkg/kubelet/apis/deviceplugin/v1alpha/api.pb.go]: Yeah, that doesn't give me very much additional confidence. It wouldn't catch something that's used even moderately rarely.

**Comment by lavalamp** [comment]
Since it's early in the cycle, and it's pretty difficult to get additional confidence, I'm inclined to take a risk and merge this. Be prepared to roll it back or fix quickly if we find problems :)

/approve

**Comment by lavalamp** [comment]
BTW, what's the performance impact of not using the unsafe package?

**Comment by jpbetz** [comment]
@wojtek-t Is there a recommended way to flag this PR for performance testing? It's a strong candidate for having noticeable performance implications.

**Comment by jpbetz** [comment]
> BTW - I didn't look carefully into the whole PR, but where exactly we stopped using unsafe package?

gogo/protobuf: https://github.com/gogo/protobuf/releases/tag/v0.5 (https://github.com/gogo/protobuf/pull/343 has benchmarks)

**Comment by wojtek-t** [comment]
Thanks - that's useful to know.

FYI: the test from yesterday is green, so it looks promissing (I will try to look deeper into individual metrics early next week).
